check_api/src/main/java/com/google/errorprone/ErrorPronePlugins.java

```diff
@@ -44,18 +47,18 @@ public static ScannerSupplier loadPlugins(ScannerSupplier scannerSupplier, Error
       loader = ErrorPronePlugins.class.getClassLoader();
     }
 
-    Iterable<BugChecker> extraBugCheckers = ServiceLoader.load(BugChecker.class, loader);
-    if (Iterables.isEmpty(extraBugCheckers)) {
-      return scannerSupplier;
-    }
+    ImmutableList<Class<? extends BugChecker>> extraBugCheckers =
+        ServiceLoader.load(BugChecker.class, loader).stream()
+            .map(ServiceLoader.Provider::type)
```

Comment:
I think this requires doing some amount of classloading, which has caused classloading issues to bubble up to javac in the past.  We should probably rework this so that we leave this as an `Iterable` and do the necessary mapping inside the code with error-handling (if it's enabled)

Reply:
Ah, ok. I was just trying to take their changes and reconcile them with what we had. I can look more into this.

Reply:
I can still update it if you like, but I think [`ServiceLoader.Provider`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.Provider.html) may be safe:

> When using a loader's [stream()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.html#stream()) method then the elements are of type Provider. This allows processing to select or filter on the provider class without instantiating the provider.

> [Class](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Class.html)<? extends [S](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.Provider.html)> type()
Returns the provider type. There is no guarantee that this type is accessible or that it has a public no-args constructor. The [get()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.Provider.html#get()) method should be used to obtain the provider instance.

Reply:
Ah gotcha, when I read the signature of the `type()` method I was thinking it loaded the underlying type and returned that rather than the provider type.  If that's the case than I think this should be safe?

Reply:
You were right the first time. I made a META-INF file with a non-existent class in it. I get `com.hubspot.TestInterfaceToLoad: Provider com.hubspot.NotARealClass not found` from
```java
List<Class<? extends TestInterfaceToLoad>> types = serviceLoader
  .stream()
  .map(Provider::type)
  .collect(Collectors.toList());
```